### ISSUE TITLE
How to load org.pepstock.charba.client.resources(containing .js) after gwt codesplit

### ISSUE THREAD

**Issue by peet-hex** [pepstock-org/Charba#27]
I have a linechart loading after  GWT.runAsync(new RunAsyncCallback()
But GWT Compile report indicates bulk of chart.js resources >200kB are loaded with initial pageload.
Is there a way to remedy this or am I missing something?

**Comment by stockiNail** [comment]
@peet-hex chart.js bundle is 205k (minify). Could you give me more info about compile report to have a look?

**Comment by peet-hex** [comment]
I take standard GreetingService example that comes with GWT.
I add charba-2.0.jar dependancy.
I add the following code to onModuleLoad method
 
 final Button graphButton = new Button("Graph");
    RootPanel.get("sendButtonContainer").add(graphButton);
    graphButton.addClickHandler(new ClickHandler() {
      @Override
      public void onClick(ClickEvent event) {
        GWT.runAsync(new RunAsyncCallback() {
          public void onFailure(Throwable caught) {
            Window.alert("Graph download failed");
          }
          public void onSuccess() {
            RootPanel.get("errorLabelContainer").add(new LineChart());
          }
        });
      }
    });
I expect to see graph added when I press the Graph button. This works as expected.
I expect the compiled code to be split in 2 sections with original 100kB segment and the new 'graphics dependant' section about 300kB but instead I find Initial load almost 400kB.
Screenshot of compile report 
![image](https://user-images.githubusercontent.com/48016440/53572953-20dbd100-3b75-11e9-9f11-a2e327573948.png)
and breakdown analysis
![image](https://user-images.githubusercontent.com/48016440/53572991-39e48200-3b75-11e9-961a-96a4167762fd.png)
How can I force org.pepstock.charba.client.resources to be loaded after the split?

**Comment by peet-hex** [comment]
Sorry its 320kB initial download instead of 100kB, but you get the point?

**Comment by stockiNail** [comment]
got it! I'm not an expert of GWT code splitting (never used) but I understood. Let me take time to have a look. I have already the project like yours and the same data from the report.

**Comment by stockiNail** [comment]
@peet-hex first of all, thank you very much for this issue. Very interesting topic.

Well, after a deep dive how the compiler is managing the splits, I found out why the initial loading was so big.

The issue was related to the dependencies. 

The class [Injector](https://github.com/pepstock-org/Charba/blob/2.1/src/org/pepstock/charba/client/Injector.java) was injecting both CHART.JS and CHARBA js module (if not already done) and this is called from constructor of classes which needs them:

```java
/**
 * Injects ChartJS and Charba utilities if not injected yet.
 */
public static void ensureInjected() {
	// use chart.js
	ensureInjected(Resources.INSTANCE.chartJs());
	// use CHARBA helper
	ensureInjected(Resources.INSTANCE.charbaHelper());
}

/**
 * Injects a script resource if not injected yet.
 * 
 * @param resource script resource
 */
public static void ensureInjected(TextResource resource) {
	// checks if already injected
	if (!ELEMENTS_INJECTED.contains(resource.getName())) {
		// creates a script element
		ScriptElement scriptElement = Document.get().createScriptElement();
		// sets ID
		scriptElement.setId(CHARBA_PREFIX_SCRIPT_ELEMENT_ID + resource.getName());
		// sets the script content source
		scriptElement.setInnerText(resource.getText());
		// appends to the body
		Document.get().getBody().appendChild(scriptElement);
		ELEMENTS_INJECTED.add(resource.getName());
	}
}
```

Nevertheless, the `Resources.INSTANCE.charbaHelper()` is needed ONLY for [JsHelper](https://github.com/pepstock-org/Charba/blob/2.1/src/org/pepstock/charba/client/commons/JsHelper.java) and it doesn't make sense to include also CHART.JS inside this helper.

Therefore, updating the [JsHelper](https://github.com/pepstock-org/Charba/blob/2.1/src/org/pepstock/charba/client/commons/JsHelper.java), as following:

```java
/**
 * To avoid any instantiation
 */
private JsHelper() {
	// to be sure that CHARBA java script object is injected
	Injector.ensureInjected(Resources.INSTANCE.charbaHelper());
}
```	

and then, using your code and compiling the project, this is the result of Compiler 

![report](https://user-images.githubusercontent.com/11741250/53630866-6a79f980-3c11-11e9-8531-7408bd22b7b0.png)


![report](https://user-images.githubusercontent.com/11741250/53630779-24bd3100-3c11-11e9-8042-f4c802fda677.png)

It sounds what you would like, doesn't it?

**Comment by peet-hex** [comment]
This sounds like what I wanted, but still questions remain:
1) When I look with a browser the size of the initial .js load still includes the bulk of the code, although the compile report indicates it should not?
2) Would it be possible to not compile in the chart.bundle.min.js as a TextResource, but keep it an external file, otherwise the compiled .js files contain 5 copies of it, which is rather undesirable in our small embedded environment.

**Comment by stockiNail** [comment]
External text resources are async therefore we must change the logic how to inject chart.js.

Let me try using Future in order to maintain a sync way to use Charba leveraging on external resources.

**Comment by stockiNail** [comment]
I have done some tests and using the external text resources you are not sure that Chart.js is injected before using it.

Due to Javascript is not multi-threading, there isn't any way to wait the load of resources (blocking is not possible).

The only way is to change completely the way to invoke Charba, and then Chart.js, having an unique entry point (resource callback) and then, when the resource is successfully loaded, start invoking the API.

Force to have an entry point is in my opinion a constraint that probably is too strong to fit your use case. 

I'm also having a look to Promise but the result sounds to be same.

**Comment by stockiNail** [comment]
@peet-hex if not urgent, let me take time to think about a possible solution

**Comment by peet-hex** [comment]
Not urgent, take your time.

**Comment by stockiNail** [comment]
@peet-hex I have created 2 interfaces which are extending the `resources` one, which contains all other resources but not CHART.JS:

 1. `EmbeddedResources` where chart.js is a textresource
 1. `ExternalResources` where chart.js is a externaltextresource

Then, before using Charba, you should set which kind of resources you want to use (and that's mandatory):

```java
DefaultsResources.setResources(ExternalResources.INSTANCE);
// or
DefaultsResources.setResources(EmbeddedResources.INSTANCE);
```
Setting the embedded one, nothing change, only the previous statement.

For users who wants to leverage on async loading fo CHART.js or code splitting, they should:

```java
GWT.runAsync(new RunAsyncCallback() {
	@Override
	public void onFailure(Throwable throwable) {
		Window.alert("Code download failed");
	}

	@Override
	public void onSuccess() {
		DefaultsResources.setResources(ExternalResources.INSTANCE);
		try {
			ExternalResources.INSTANCE.chartJs().getText(new ResourceCallback<TextResource>() {
					
				@Override
				public void onSuccess(TextResource resource) {
					Injector.ensureInjected(resource);
					Defaults.get();
					RootPanel.get().add(new LineChart());
				}
							
				@Override
				public void onError(ResourceException e) {
					Window.alert("Unablke to load chart.js");
				}
			});
		} catch (ResourceException e) {
			e.printStackTrace();
		}
	}
});
```

This is the result of Compiler report:

<img width="331" alt="report2" src="https://user-images.githubusercontent.com/11741250/53842809-d4432c00-3fa0-11e9-9f92-6ba0d388d328.png">


The code splitting file are all 93k.

I haven't committed yet. I'm still thinking how I can improve it or name of classes.

What do u think?

**Comment by peet-hex** [comment]
This looks like a it can really solve the problem! 
But I see what you mean to improve in terms of complexity. Maybe find a way to refactor the ResourceCallback back into one of the Resources classes and hide it away from the end user? But then you still have to test for a null return value or similar?
The compile report looks promising, but If you monitor the initial .js file download in chrome developer tools, what is the size then?

**Comment by stockiNail** [comment]
@peet-hex what I put here was just the code for 'testing'. Implementing a ResourceCallback, setting the resource type, is something we will do.
Nevertheless you got my point. For whom wants to use Charba (as is today) without splitting, she/he adds the mandatory set for resource type... And it could be annoying.

And I can not use any default, because setting default (the embedded one as is today) the compiler didn't split the client bundle and you will have the chart.js source to download anyway.

My idea is to implement it anyway, change a little names, implementing ResourceCallback and also checking the null value, throwing an exception.

When the updates will be committed, you could see them linked here

**Comment by stockiNail** [comment]
@peet-hex About the size of initial.js, I'm a little surprised because seeing into the browser, JS file is 211k (initial size + full size). Honestly having a look to the compile report, the picture is completely different, as you wrote.

**Comment by stockiNail** [comment]
Committed the first implementation.

I have created 2 interfaces which are extending the `resources` one, which contains all other resources but not CHART.JS:

 1. `EmbeddedResources` where chart.js is a textresource
 1. `DeferredResources` where chart.js is a externaltextresource

Then, before using Charba, you should set which kind of resources you want to use (and that's mandatory):

```java
ResourcesType.setClientBundle(EmbeddedResources.INSTANCE);
// or
ResourcesType.setClientBundle(DeferredResources.INSTANCE);
```
Setting the embedded one, nothing change, only the previous statement.

For users who wants to leverage on async loading of CHART.js or code splitting, they should:

```java
GWT.runAsync(new RunAsyncCallback() {
	@Override
	public void onFailure(Throwable throwable) {
		Window.alert("Code download failed");
	}

	@Override
	public void onSuccess() {
		EntryPointStarter.run(new Runnable() {
			
			@Override
			public void run() {
				// UI cretion and starting point to invoke CHARBA API
			}
		});
	}
});
```
About the previous post, I think that's correct how is working. 

Let me know what u think about.

**Comment by stockiNail** [comment]
@peet-hex  released version [2.2](https://github.com/pepstock-org/Charba/releases/tag/2.2) with the implementation described above
